zarr/codecs.py

```diff
@@ -485,6 +485,100 @@ def __repr__(self):
 codec_registry[Delta.codec_id] = Delta
 
 
+class AsType(Codec):
+    """Filter to convert data between different types.
+
+    Parameters
+    ----------
+    encode_dtype : dtype
+        Data type used to store the data.
+    decode_dtype : dtype, optional
+        Data type of the data read in.
+
+    Notes
+    -----
+    If `encode_dtype` is of lower precision than `decode_dtype`, please be
+    aware that data loss can occur by writing data to disk using this filter.
+    No checks are made to ensure the casting will work in that direction and
+    data corruption will occur.
+
+    Examples
+    --------
+    >>> import zarr
+    >>> import numpy as np
+    >>> x = np.arange(100, 120, 2, dtype='i1')
+    >>> x
+    array([100, 102, 104, 106, 108, 110, 112, 114, 116, 118], dtype=int8)
+    >>> f = zarr.AsType(encode_dtype=x.dtype, decode_dtype='i8')
+    >>> y = f.decode(x)
+    >>> y
+    array([100, 102, 104, 106, 108, 110, 112, 114, 116, 118])
+    >>> z = f.encode(y)
+    >>> z
+    array([100, 102, 104, 106, 108, 110, 112, 114, 116, 118], dtype=int8)
+
+    """  # flake8: noqa
+
+    codec_id = 'astype'
+
+    def __init__(self, encode_dtype, decode_dtype):
+        self.encode_dtype = np.dtype(encode_dtype)
+        self.decode_dtype = np.dtype(decode_dtype)
+
+    def encode(self, buf):
+
+        # view input data as 1D array
+        arr = _ndarray_from_buffer(buf, self.decode_dtype)
+
+        # setup encoded output
+        enc = np.empty_like(arr, dtype=self.encode_dtype)
+
+        # Convert and copy
+        enc[...] = arr.astype(self.encode_dtype)
+
+        return enc
+
+    def decode(self, buf, out=None):
+
+        # view encoded data as 1D array
+        enc = _ndarray_from_buffer(buf, self.encode_dtype)
+
+        # setup decoded output
+        if isinstance(out, np.ndarray):
+            # optimization, can decode directly to out
+            dec = out.reshape(-1, order='A')
+            copy_needed = False
+        else:
+            dec = np.empty_like(enc, dtype=self.decode_dtype)
+            copy_needed = True
+
+        # Convert and copy
+        dec[...] = enc.astype(self.decode_dtype)
+
+        # handle output
+        if copy_needed:
+            out = _buffer_copy(dec, out)
```

Comment:
Apologies, the logic is a bit tortuous here, I'm having trouble myself reconstructing everything after not looking at it for a while. I *think* this can actually be simplified as follows (which will also remove a memory copy in the case where out is None):

```
        # view encoded data as 1D array
        enc = _ndarray_from_buffer(buf, self.encode_dtype)

        # apply decoding
        dec = enc.astype(self.decode_dtype)

        # handle output
        out = _buffer_copy(dec, out)
```